File: backend/schemas/history.py
"""History response schemas (T-306, RF-050–051, UC-050–051)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from schemas.prediction import PredictRequest, RiskLevel, ShapExplanationItem


class HistoryPatientSummary(BaseModel):
    """De-identified patient snapshot for history list rows."""

    age: int | None = None
    gender: str | None = None
    glucose: float | None = None
    hospital_stay_days: int | None = None


class HistoryPatientDetail(BaseModel):
    """Persisted clinical inputs for a historical evaluation (RF-052)."""

    age: int | None = None
    gender: str | None = None
    glucose: float | None = None
    blood_pressure: float | None = None
    medications_count: int | None = None
    previous_admissions: int | None = None
    hospital_stay_days: int | None = None
    bmi: float | None = None


class HistorySimulationItem(BaseModel):
    """Stored what-if simulation linked to a prediction (UC-052)."""

    id: UUID
    created_at: datetime
    original_risk_percent: float = Field(ge=0.0, le=100.0)
    simulated_risk_percent: float = Field(ge=0.0, le=100.0)
    delta_risk_percent: float
    simulation_summary: str | None = None


class HistoryUserSummary(BaseModel):
    """Clinician who performed the evaluation."""

    id: UUID
    email: str
    first_name: str
    last_name: str
    role: str


class HistoryListItem(BaseModel):
    """Single prediction row in history (UC-050)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    risk_score: float = Field(ge=0.0, le=1.0)
    risk_percent: float = Field(ge=0.0, le=100.0)
    risk_level: RiskLevel
    confidence_score: float | None = Field(default=None, ge=0.0, le=1.0)
    summary: str | None = None
    model_version: str
    prediction_time_ms: int | None = None
    created_at: datetime
    user: HistoryUserSummary
    patient_input: HistoryPatientSummary | None = None


class HistoryListResponse(BaseModel):
    """Paginated prediction history (RF-051)."""

    items: list[HistoryListItem]
    total: int
    limit: int
    offset: int


class HistoryDetailResponse(BaseModel):
    """Historical prediction detail with inputs, SHAP, and simulations (RF-052, UC-052)."""

    id: UUID
    risk_score: float = Field(ge=0.0, le=1.0)
    risk_percent: float = Field(ge=0.0, le=100.0)
    risk_level: RiskLevel
    confidence_score: float | None = Field(default=None, ge=0.0, le=1.0)
    summary: str | None = None
    model_version: str
    prediction_time_ms: int | None = None
    created_at: datetime
    user: HistoryUserSummary
    patient_input: HistoryPatientDetail | None = None
    baseline_request: PredictRequest
    shap_explanations: list[ShapExplanationItem]
    simulations: list[HistorySimulationItem]

File: frontend/src/types/history.ts
/** Prediction history API types (aligned with backend/schemas/history.py). */

import type { PredictRequest, RiskLevel, ShapExplanationItem } from "@/types/prediction";

export interface HistoryPatientSummary {
  age: number | null;
  gender: string | null;
  glucose: number | null;
  hospital_stay_days: number | null;
}

export interface HistoryPatientDetail {
  age: number | null;
  gender: string | null;
  glucose: number | null;
  blood_pressure: number | null;
  medications_count: number | null;
  previous_admissions: number | null;
  hospital_stay_days: number | null;
  bmi: number | null;
}

export interface HistorySimulationItem {
  id: string;
  created_at: string;
  original_risk_percent: number;
  simulated_risk_percent: number;
  delta_risk_percent: number;
  simulation_summary: string | null;
}

export interface HistoryUserSummary {
  id: string;
  email: string;
  first_name: string;
  last_name: string;
  role: string;
}

/** Single prediction row in GET /history (UC-050). */
export interface HistoryListItem {
  id: string;
  risk_score: number;
  risk_percent: number;
  risk_level: RiskLevel;
  confidence_score: number | null;
  summary: string | null;
  model_version: string;
  prediction_time_ms: number | null;
  created_at: string;
  user: HistoryUserSummary;
  patient_input: HistoryPatientSummary | null;
}

/** Paginated history list (RF-051). */
export interface HistoryListResponse {
  items: HistoryListItem[];
  total: number;
  limit: number;
  offset: number;
}

/** Query params for GET /history (UC-051). */
export interface HistoryListParams {
  risk_level?: RiskLevel;
  user_id?: string;
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
}

/** Historical prediction detail (RF-052, UC-052). */
export interface HistoryDetailResponse {
  id: string;
  risk_score: number;
  risk_percent: number;
  risk_level: RiskLevel;
  confidence_score: number | null;
  summary: string | null;
  model_version: string;
  prediction_time_ms: number | null;
  created_at: string;
  user: HistoryUserSummary;
  patient_input: HistoryPatientDetail | null;
  baseline_request: PredictRequest;
  shap_explanations: ShapExplanationItem[];
  simulations: HistorySimulationItem[];
}
